app/portainer/home/EnvironmentList/KubeconfigButton.tsx

```diff
@@ -34,48 +81,156 @@ export function KubeconfigButton({ environments }: Props) {
       category: 'kubernetes',
     });
 
-    showKubeconfigModal(environments);
+    fetchKsExpiry();
+    selection.clear();
+    setShowDialog(true);
+  }
+
+  function handleClose() {
+    setShowDialog(false);
+  }
+
+  function handleDownload() {
+    confirmKubeconfigSelection();
+  }
+
+  function isKubeconfigButtonVisible(environments: Environment[]) {
+    if (window.location.protocol !== 'https:') {
+      return false;
+    }
+    return environments.some((env) => isKubernetesEnvironment(env.Type));
+  }
+
+  async function fetchKsExpiry() {
+    let expiryMessage = '';
+    try {
+      expiryMessage = await kcService.expiryMessage();
+    } catch (e) {
+      notifications.error('Failed fetching kubeconfig expiry time', e as Error);
+    }
+    setKsExpiry(expiryMessage);
   }
-}
 
-function isKubeconfigButtonVisible(environments: Environment[]) {
-  if (window.location.protocol !== 'https:') {
-    return false;
+  async function confirmKubeconfigSelection() {
+    if (selection.size === 0) {
+      notifications.warning('No environment was selected', '');
+      return;
+    }
+    try {
+      await kcService.downloadKubeconfigFile(Array.from(selection));
+      setShowDialog(false);
+    } catch (e) {
+      notifications.error('Failed downloading kubeconfig file', e as Error);
+    }
   }
-  return environments.some((env) => isKubernetesEnvironment(env.Type));
 }
 
-async function showKubeconfigModal(environments: Environment[]) {
-  const kubeEnvironments = environments.filter((env) =>
+export interface KubeconfigPormptProps {
+  ksExpiry: string;
+  selection: Set<number>;
+  envQueryParams: EnvironmentsQueryParams;
+}
+const storageKey = 'home_endpoints';
+
+export function KubeconfigPormpt({
+  ksExpiry,
+  selection,
+  envQueryParams,
+}: KubeconfigPormptProps) {
+  const [page, setPage] = useState(1);
+  const [pageLimit, setPageLimit] = usePaginationLimitState(storageKey);
+  const [checked, setChecked] = useState(false);
+  const [checkAll, setCheckAll] = useState(false);
+  const { environments, totalCount } = useEnvironmentList(
+    { page, pageLimit, ...envQueryParams },
+    true
+  );
+  const kubeEnvs = environments.filter((env) =>
```

Comment:
one of the `useEnvironmentList` params is `types` where you can query only by `kubernetes` types

Reply:
because you don't use this , totalCount isn't correct (as it shows the total count for all environments, not just kube)

Reply:
@chaogeng77977 please use this